### ISSUE TITLE
Question: the tools incorrectly rotates the image?

### ISSUE THREAD

**Issue by MoAbbasid** [CVHub520/X-AnyLabeling#447]
Hi, The tool shows the image in wrong orientation, Maybe has issues with EXIF data, but when I use Anylabeling tool which is similar to this one the image loads correctly, how to solve that?
![image](https://github.com/CVHub520/X-AnyLabeling/assets/109457597/4cafc6f0-70f2-4cb2-9e36-9689e1745c10)

**Comment by CVHub520** [comment]
Hi, there!

Currently, you can simply check the current image's EXIF information with the following script:

```python
from PIL import Image, ExifTags

def get_exif_data(image_path):
    try:
        with Image.open(image_path) as img:
            exif_data = img._getexif()
            if exif_data is not None:
                print("EXIF Information:")
                for tag, value in exif_data.items():
                    tag_name = ExifTags.TAGS.get(tag, tag)
                    print(f"{tag_name}: {value}")
            else:
                print("No EXIF information found.")
    except Exception as e:
        print(f"Error: {e}")

image_path = "cat_dataset/images/IMG_20210627_225110.jpg"


get_exif_data(image_path)
```

And then, as an auxiliary solution, you can simply use the following script to convert the images to a normal state:

```python
from PIL import Image, ExifTags
import os

def rotate_images_in_folder(folder_path):
    for filename in os.listdir(folder_path):
        if filename.endswith(('.jpg', '.jpeg', '.png')):
            image_path = os.path.join(folder_path, filename)
            rotate_image(image_path)

def rotate_image(image_path):
    try:
        with Image.open(image_path) as img:
            exif_data = img._getexif()
            for tag, value in exif_data.items():
                tag_name = ExifTags.TAGS.get(tag, tag)
                if tag_name != "Orientation":
                    continue
                if value == 3:
                    img = img.rotate(180, expand=True)
                elif value == 6:
                    img = img.rotate(270, expand=True)
                elif value == 8:
                    img = img.rotate(90, expand=True)
                img.save(image_path)
                print(f"Image {image_path} rotated successfully.")
            else:
                print(f"No rotation needed for {image_path}.")
    except Exception as e:
        print(f"Error processing {image_path}: {e}")

folder_path = "/home/cvhub/workspace/tmp/20240118/cat/cat_dataset/images"
rotate_images_in_folder(folder_path)
```

Finally, re-import into the tool for annotation.

Good luck!

#216

**Comment by MoAbbasid** [comment]
Thank you, the provided code works and now the Images are rotated correctly, another question, is it possible to set train/val splits from the gui? and are there docs on how to use the cli mode?

**Comment by CVHub520** [comment]
You're welcome! We're glad to hear that the provided code resolved the issue and that your images are now rotated correctly. 🤗

Regarding the train/val splits, this functionality typically falls under the purview of the training platform rather than the annotation tool. Since different training frameworks may require different formats for the splits, we currently do not offer a unified interface for this within X-AnyLabeling.

However, if you're interested in running X-AnyLabeling from the command line, we do have documentation available to guide you through the process. You can find the documentation here: [get_started](https://github.com/CVHub520/X-AnyLabeling/blob/main/docs/en/get_started.md).

If you have any more questions or need further assistance, feel free to reach out. We're here to help! 😊

Best regards,
CVHub

**Comment by MoAbbasid** [comment]
@CVHub520 , Hi again
I'm trying to get some of the team members to switch to your tool instead of other tools, but this issue of rotating images keeps popping up and only in this tool, The code you provided to fix the rotation works but its an extra step I have to teach them before using the tool, and some of them are not very techsavy.

my suggestion since its seems like an easy fix, maybe its worth it to implement it and fix it?
why does this issue only appear here and not in labelme or anylabeling for example? 

Thank you

**Comment by CVHub520** [comment]
Hello,

Thank you for reaching out and for your suggestion. I understand the inconvenience caused by the image rotation issue and appreciate your effort in advocating for the tool among your team members.

The reason this issue might be specific to our tool could be due to differences in how image metadata is interpreted or handled during the loading process. 

I want to assure you that I am committed to improving the tool, and I will prioritize addressing this issue. I agree that it would be beneficial to integrate a fix directly into the tool to streamline the user experience, especially for those who may not be as technically inclined.

I will work on implementing a solution and will keep you updated on the progress. Once the fix is ready, I will release an update, and I hope this will make it even easier for your team to adopt the tool.

Thank you for your patience and support. I'm here to help with any further questions or concerns you may have.

Best regards,
CVHub 🌟

**Comment by CVHub520** [comment]
@MoAbbasid, Hi there!

I hope this message finds you well. I'm thrilled to let you know that the image rotation issue you encountered has now been addressed. I've implemented a fix that should streamline the process for all users, including those who may not be as technically inclined.

Please update to the latest version of our tool to experience the improvements. We appreciate your patience and your suggestion, which has helped us enhance our product.

If you have any further feedback or questions, feel free to reach out. We're here to support you every step of the way.

Best regards,
CVHub

#198, #216, #291

**Comment by chenller** [comment]
May I ask if the fix for this bug has been merged into the main branch? Do I still need to run a repair script for the images annotated on the latest main branch code?

**Comment by CVHub520** [comment]
> May I ask if the fix for this bug has been merged into the main branch? Do I still need to run a repair script for the images annotated on the latest main branch code?

Sure, the bug fix you're referring to has indeed been integrated into the main branch. As for the images annotated using the latest main branch code, it's no longer necessary to run a repair script. For more details, feel free to check out [this line](https://github.com/CVHub520/X-AnyLabeling/blob/3dca8cad84e1f1a44027769f798187ff168c12c5/anylabeling/views/labeling/label_file.py#L45), where there are no difference between these methods.

**Comment by chenller** [comment]
This is a great solution. This is very useful.
Additionally, based on the information I have searched for, the following code can eliminate the rotation caused by the code `Image.open(file_name)` for your reference.
```python
from PIL import Image, ImageOps

img = Image.open(str(image_path))
img = ImageOps.exif_transpose(img)  # Restore images at normal angles
```



> > May I ask if the fix for this bug has been merged into the main branch? Do I still need to run a repair script for the images annotated on the latest main branch code?
> 
> Sure, the bug fix you're referring to has indeed been integrated into the main branch. As for the images annotated using the latest main branch code, it's no longer necessary to run a repair script. For more details, feel free to check out [this line]
(https://github.com/CVHub520/X-AnyLabeling/blob/3dca8cad84e1f1a44027769f798187ff168c12c5/anylabeling/views/labeling/label_file.py#L45), where there are no difference between these methods.

**Comment by CVHub520** [comment]
> This is a great solution. This is very useful. Additionally, based on the information I have searched for, the following code can eliminate the rotation caused by the code `Image.open(file_name)` for your reference.
> 
> ```python
> from PIL import Image, ImageOps
> 
> img = Image.open(str(image_path))
> img = ImageOps.exif_transpose(img)  # Restore images at normal angles
> ```
> 
> > > May I ask if the fix for this bug has been merged into the main branch? Do I still need to run a repair script for the images annotated on the latest main branch code?
> > 
> > 
> > Sure, the bug fix you're referring to has indeed been integrated into the main branch. As for the images annotated using the latest main branch code, it's no longer necessary to run a repair script. For more details, feel free to check out [this line]
> > (https://github.com/CVHub520/X-AnyLabeling/blob/3dca8cad84e1f1a44027769f798187ff168c12c5/anylabeling/views/labeling/label_file.py#L45
> > ), where there are no difference between these methods.

Woo! Thank you for your provided solution, it looks more straightforward! 😄

**Comment by MoAbbasid** [comment]
@chenller wow that one line works and its indeed much more concise, thanks for that

**Comment by CVHub520** [comment]
Q1: https://www.pythonguis.com/installation/install-pyqt-mac/
Q2: Not implemented yet.
